src/porepy/numerics/ad/operators.py

```diff
@@ -454,31 +420,63 @@ def error_message(operation):
                     "Encountered a case not covered when dividing Ad objects"
                 )
 
+        elif tree.op == Operation.evaluate:
+            # This is a function, which should have at least one argument
+            assert len(results) > 1
+            return results[0].func(*results[1:])
+
         else:
             raise ValueError("Should not happen")
 
-    def _parse_readable(self, op: "Operator") -> str:
-        # Make a human-readable error message related to a parsing error.
-        # NOTE: The exact formatting should be considered work in progress,
-        # in particular when it comes to function evaluation.
+    def _get_error_message(self, operation: str, tree, results: list) -> str:
+        # Helper function to format error message
+        msg_0 = tree.children[0]._parse_readable()
+        msg_1 = tree.children[1]._parse_readable()
+
+        nl = "\n"
+        msg = (
+            f"Ad parsing: Error when {operation}\n"
+            + "  "
+            + msg_0
+            + nl
+            + "with"
+            + nl
+            + "  "
+            + msg_1
+            + nl
+        )
+
+        msg += (
+            f"Matrix sizes are {_get_shape(results[0])} and "
+            f"{_get_shape(results[1])}"
+        )
+        return msg
+
+    def _parse_readable(self) -> str:
+        """
+        Make a human-readable error message related to a parsing error.
+        NOTE: The exact formatting should be considered work in progress,
+        in particular when it comes to function evaluation.
+        """
 
         # There are three cases to consider: Either the operator is a leaf,
         # it is a composite operator with a name, or it is a general composite
         # operator.
-        if op.is_leaf():
+        if self.is_leaf():
```

Comment:
I agree, changing to `self`, and calling this function on `tree.children `(above) makes more sense.